### ISSUE TITLE
Future of `vio` checking

### ISSUE THREAD

**Issue by ejgallego** [coq/coq#8767]
We were discussing about cleanups of `library` and it came to my mind the question of what is the future of non-checking compilation in Coq.

I think that the `.vio` / `-quick` stuff is not very much used in the end, in particular due to problems with file size. So maybe we should deprecate it and move to proposals such as #8642 ?

cc: @gares

**Comment by ejgallego** [comment]
> Why do you say it is not used?

Does for example any of the CI developments use it? [I will check]

It seems to me tho that the current functionality doesn't match well with what quite a few users want, based on personal discussions.

Indeed there is no point in deprecating it until we have something ready, my question is precisely about what the plans are? Should we embrace a different method and drop the `vio -> vo` feature? Should we instead fix the problems with `vio`?

It seems to me that having a plan / roadmap before doing the work itself can be useful?

**Comment by gares** [comment]
> Does for example any of the CI developments use it? [I will check]

How can you check that? The best you can do is to ask people if they ever type `make quick` in their daily workflow. Given the level of parallelism we have on CI I doubt if would be of any use to to a `v -> vio -> vo` compilation path, since the latter is a gain only if -j can be high.

The size problem of vio files is something only @ppedrot really investigated. Maybe he can post his findings, eg if it is fixable or not.

**Comment by gares** [comment]
I won't be there. Not sure I agree with you about the code complexity. Please ask PMP about the size issue

**Comment by ejgallego** [comment]
I had indeed seen in the calendar that you won't be able to attend, that's why I will try to write a summary of the discussion here. Arthur added to the WG program a discussion on his PR so I guess we will take that slot to see how could we move forward.

**Comment by ejgallego** [comment]
> Not sure I agree with you about the code complexity.

I was curious so I just submitted an experiment at https://github.com/coq/coq/pull/8783 This should also help understanding CI impact.

IMHO the complexity is more than the gain, in particular due to the effect on lower layers of the kernel, but I guess I still want to understand the code better.

**Comment by ejgallego** [comment]
We had some quite long discussion in the WG, I feel unable to properly summarize it, so help from others @charguer @ppedrot @maximedenes is welcome. Some points that were made:

- @ejgallego : "the current approach replaces a proof term with something that is O(|state|), thus `.v -> .vio -> .vo` is going to be worse in most cases than `.v -> .vo`.
- @ppedrot : "I think it is easy to store only the diff of the state, that problem would still be diminished"
- @ejgallego : "I am surprised you think this is easy, I was under the impression that would be difficult, so indeed, if you think so why not. Even so in general I am not convinced, in the sense that when a proof is not there I'd rather have a full hole and re-generate later"

Some more questions on the real-world use of `.vio`.
- @ejgallego : "Not clear if people is using that, reports is that it was too slow"
- @ppedrot : "I suggest we write to coq-club to ask users"
  @gares , are you OK if we write to coq-club to gather user feedback?

Some discussion on what to do, options are:
- keep vio + @charguer proposal.
- keep only improved vio.
- keep only @charguer proposal,
- develop a new system.
- @maximedenes says that keeping both make sense if they have different functionality.
- @ejgallego says that in practice he believes they do. Some disagreement about this.

Then @charguer joined and commented on his proposal.

**Comment by gares** [comment]
>  @gares , are you OK if we write to coq-club to gather user feedback?

yes

I had some private discussions with @charguer  before he opened his PR.
I believe his approach is fine when Require is fast. In this case rebuilding the state one needs in order to process the proofs is cheap. If Require is slow, then unmarshaling the state is faster. I'm over simplifying a bit here, but I believe the two approaches cover two similar but not equal scenarios.
I'd be happy to be proved wrong and replace .vio with this other approach. As far as I can tell, the code complexity and limitations of the two are pretty close.

**Comment by ejgallego** [comment]
> As far as I can tell, the code complexity and limitations of the two are pretty close.

Oh interesting you say that, in an informal analysis I got the impression code complexity differed quite a bit, I'll do a formal analysis when I get a minute.

**Comment by charguer** [comment]
Following up on this discussion, my understanding is that the only case where the .vio approach would have a serious benefit is if the time for executing the `Require` commands is significantly longer than loading the `.vio` that was built from marshalling a state that previously executed these commands.

I don't know the internals of the implementation, but it seems to me that a `Require` should do little more than unmarshalling big tables of definitions and add them to the current state (at a high level). Thus, if a `Require` is slower than loading a `vio`, then perhaps it means that there is something suboptimal in the way `Require` is handled.

If we have good reason to think that `vio` can speedup significantly the `Require` and that this is of key importance for a number of users, then we should probably keep `vio` alive. Otherwise, I suggest we should mark the `.vio` approach as deprecated and remove it when the `-vos` approach becomes mature enough.

In any case, the first step is to get feedback from current `vio` users by contacting them on coq-club, since these users are the only ones concerned by the potential drop of `vio` files.

**Comment by gares** [comment]
I usually speak about `Require` because it is a slow and easy to explain part of the picture.
Let me try with a slightly more complex example:
```
Require foo.
Lemma bar : T. Proof ... Qed.
Definition bad : T. Proof. tactics. Defined.
Lemma baz : T. Proof. ... Qed.
```

The .vio would contain a dump of the system state at the beginning of `bar` and `baz`, along with the ommitted proof text to be executed in these states. `bad` is executed only in `v -> vio`.

If I understand correctly the other approach, in order to check `bar` one needs to replay `Require`, but in order to check `baz` one would need to execute `bad` (or any other Definition, Notation, Instance).

Not that I like when Definitions are used as proofs, but there are heavy users of this style, and they do so because, for example, in there they can call `abstract ... as name` and use `name` in the proof of `baz`.
These users are also the ones that complained about .vio size. For them, the alternative approach would be even more problematic, if I'm not mistaken.

**Comment by gares** [comment]
The point I wanted to make is that, if a file spends 99% of the time to do the opaque proofs and 1% on the rest, the new vos/vok approach seems better to me. If the proportion is less biased, then vio may still be the best one can get.

**Comment by SkySkimmer** [comment]
>bad is executed only in v -> vio.

What? That's not actually true is it?

**Comment by gares** [comment]
Of course it is, why would I lie? The task saved on disk are only the ones for the opaque proofs, all the rest is executed once.

**Comment by SkySkimmer** [comment]
Oh somehow I misread `v -> vio` into `vio -> vo`...
Sorry for the confusion.

**Comment by charguer** [comment]
Talking about percentages, no need to go as extreme as 99%. Let me try to prove (not in Coq) that, for a formal development whose compilation spends 2/3 of its time in proofs, the `vos` approach is better. (I suspect that few formal developments go below this ratio of 2/3).

Let's put aside the quadratic blowup problem, and just pretend, with rough figures, that overheads associated with increased disk traffic and massive serialization (due to huge vio files) incurs a 5% overheads, and managing/scheduling futures in `vio` files account for another 5% overhead in execution time.

- For working on one particular file in the development, building `vos` files gets you in a ready state 3 times faster compared with building `vo` for the dependencies. In comparison, building `vio` files can only be slower than building `vos` files. Based on the two reasons cited above, it could be ~10% slower.

- For compiling a full development, building all `vos+vok` files costs 33% more than building  `vo` files directly (since we assume less than 1/3 of the time is devoted to non-proofs). With our 10% assumed penatly for `vio` files, building all `vos+vok` files would then be only 23% slower than building all `vio+vo` files. 

In summary, for the frequent operation of getting a file into a ready state, we gain 10%. For the much less frequent operation of compiling everything, we loose 23%. In passing, we get rid of the quadratic size blowup problem. Of course, you could argue with my somewhat arbitrary figures. But I think that overall we're better off with the `vos`.

If you then consider a development with more than 5/6 of the time spend in proofs, then using my figures you'll obtain that the `vos` approach is systematically better, regardless of whether you want to compile just one file or want to compile all files.

**Comment by maximedenes** [comment]
The Working Group seems to converge towards introduce `.vos` and keeping `.vio` files as they are for now, so closing meanwhile.
